### PR TITLE
Fix monitor for TiKV-Importer

### LINKED ISSUE #3618: TiKV-Importer supports to be pulled metrics directly without Pushgateway
## Feature Request

**Is your feature request related to a problem? Please describe:**

TiKV-Importer has `status-server-address` config item for metrics pull, then the TiKV-Importer Helm chart can remove Pushgateway to make it simpler.

**Describe the feature you'd like:**

TiKV-Importer supports to be pulled metrics directly without Pushgateway

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### LINKED ISSUE #3628: Prometheus can't scrape TiDB-Lightning's metrics with the default values.yaml
## Bug Report

**What version of Kubernetes are you using?**

**What version of TiDB Operator are you using?**

v1.1.7

**What storage classes exist in the Kubernetes cluster and what are used for PD/TiKV pods?**

**What's the status of the TiDB cluster pods?**

**What did you do?**

query TiDB-Lighting's metrics in Prometheus with TiDB-Lighting installed with the default `values.yaml`.

**What did you expect to see?**

metrics can be seen in Prometheus.

**What did you see instead?**

scrape metrics timeout.

this is caused by:

- default `values.yaml` set `service.type: NodePort` in https://github.com/pingcap/tidb-operator/blob/ba2f64ce6e26a151f7583fc955429390fd2ad51d/charts/tidb-lightning/values.yaml#L13
- but Prometheus try to scrape metrics via `{service-name}.{namespace}:8289/metrics`

### PR DESCRIPTION
**Pull request by csuzhangxc**
### What problem does this PR solve?

- fix #3618

### What is changed and how does it work?

- remove Prometheus Pushgateway for TiKV-Importer and enable status server
- fix labels for TiKV-Importer so that Prometheus can re-label `__address__` correctly

some useful links:

- [`kubernetes_sd_config`](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#kubernetes_sd_config)

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [ ] Manual test 
  1. `helm install` TiKV-Importer and TiDB-Lightning
  2. import some data via TiDB-Lighting
  3. query metrics in Prometheus for TiDB-Lighting and TiKV-Importer
    
image

lightning-metrics

importer-metrics
image

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [x] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Fix the issue that Prometheus can't pull metrics for TiKV-Importer
```

### CODE REVIEW

File: charts/tikv-importer/templates/tikv-importer-service.yaml [hunk 1]
```diff
@@ -3,17 +3,18 @@ kind: Service
 metadata:
   name: {{ .Values.clusterName }}-importer
```
Comment (DanielZhangQD) [inline charts/tikv-importer/templates/tikv-importer-service.yaml:4]: I think we only need to change the service name from `{{ .Values.clusterName }}-importer` to `{{ .Release.Name }}-importer`, no need to change the other labels or selectors, please help confirm.

Comment (csuzhangxc) [reply charts/tikv-importer/templates/tikv-importer-service.yaml:4]: we use `__meta_kubernetes_pod_label_app_kubernetes_io_name` (and some other labels and annotations) in the Prometheus config. so we need to change the labels.

File: charts/tidb-lightning/values.yaml [hunk 2]
```diff
@@ -9,8 +9,8 @@ image: pingcap/tidb-lightning:v4.0.9
 imagePullPolicy: IfNotPresent
 # imagePullSecrets: []
 
-service:
-  type: NodePort
+service: {}
```
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply charts/tidb-lightning/values.yaml:12]: we use `__meta_kubernetes_pod_annotation_%s_prometheus_io_port` in the Prometheus config. If we set `type: NodePort`, then this automatically created node port can not be re-label with any Prometheus _meta labels_.

File: charts/tikv-importer/templates/tikv-importer-service.yaml (hunk 1 above)
Comment (mikechengwei) [reply charts/tikv-importer/templates/tikv-importer-service.yaml:4]: I don’t think we need to change the existing label of helm, we can add a cluster label. `app.kubernetes.io/name` and `app.kubernetes.io/instance` represents the container iteself.

**Comment by dragonly** [comment]
LGTM

**Review by dragonly — APPROVED** [review]

File: charts/tidb-lightning/values.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (DanielZhangQD) [reply charts/tidb-lightning/values.yaml:12]: Prometheus uses the service name to access the lightning service, the service name + service port can be accessed no matter the service type is nodePort or service.

File: charts/tikv-importer/templates/tikv-importer-service.yaml (hunk 1 above)
Comment (DanielZhangQD) [reply charts/tikv-importer/templates/tikv-importer-service.yaml:4]: I think the issue here is that the value of instance label is `app.kubernetes.io/instance: {{ .Release.Name }}`, which is used to construct the service name in TidbMonitor controller, however, the service name deployed by the chart is `{{ .Values.clusterName }}-importer`, and they do not match.

File: charts/tidb-lightning/values.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (csuzhangxc) [reply charts/tidb-lightning/values.yaml]: sorry, my mistake, have reverted it back.

File: charts/tikv-importer/templates/tikv-importer-service.yaml (hunk 1 above)
Comment (mikechengwei) [reply charts/tikv-importer/templates/tikv-importer-service.yaml:4]: If we not spec clustername fied , it will use release name , right? How to deal with it?

Comment (csuzhangxc) [reply charts/tikv-importer/templates/tikv-importer-service.yaml:4]: > If we don't spec clustername field , it will use release name , right?

NO, and I have removed `if not specified, the chart release name will be used` in this PR.

if don't specify this filed, an error will be reported, like:

```
Error: unable to build kubernetes objects from release manifest: error validating "": error validating data: unknown object type "nil" in ConfigMap.metadata.labels.app.kubernetes.io/instance
```

File: pkg/monitor/monitor/template.go [hunk 3]
```diff
@@ -240,11 +240,11 @@ func buildAddressRelabelConfigByComponent(kind string) *config.RelabelConfig {
 		return &config.RelabelConfig{
 			Action:      config.RelabelReplace,
 			Regex:       addressPattern,
-			Replacement: "$1.$2-importer.$3:$4",
+			Replacement: "$2.$3:$4",
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/template.go:243]: Please revert the change to this file, after changing the instance label, this should work.

File: charts/tikv-importer/templates/tikv-importer-configmap.yaml [hunk 4]
```diff
@@ -3,9 +3,9 @@ kind: ConfigMap
 metadata:
   name: {{ .Values.clusterName }}-importer-{{ template "importer-configmap.data-digest" . }}
   labels:
-    app.kubernetes.io/name: {{ template "chart.name" . }}
+    app.kubernetes.io/name: {{ .Values.clusterName }}-importer
```
Comment (DanielZhangQD) [inline charts/tikv-importer/templates/tikv-importer-configmap.yaml:6]: I think we do not need to change this, just change the instance label is OK.

File: pkg/monitor/monitor/template.go (hunk 3 above)
Comment (csuzhangxc) [reply pkg/monitor/monitor/template.go]: reverted

File: charts/tikv-importer/templates/tikv-importer-configmap.yaml (hunk 4 above)
Comment (csuzhangxc) [reply charts/tikv-importer/templates/tikv-importer-configmap.yaml]: reverted

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Comment by DanielZhangQD** [comment]
@csuzhangxc If users deploy tikv-importer of v1.1.8, can they upgrade to the chart in this PR?

**Comment by csuzhangxc** [comment]
> @csuzhangxc If users deploy tikv-importer of v1.1.8, can they upgrade to the chart in this PR?

```
Error: UPGRADE FAILED: cannot patch "basic-importer" with kind Service: Service "basic-importer" is invalid: spec.clusterIP: Invalid value: "": field is immutable && cannot patch "basic-importer" with kind StatefulSet: StatefulSet.apps "basic-importer" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden
```

File: charts/tikv-importer/templates/tikv-importer-service.yaml [hunk 5]
```diff
@@ -5,15 +5,16 @@ metadata:
   labels:
     app.kubernetes.io/name: {{ template "chart.name" . }}
     app.kubernetes.io/managed-by: {{ .Release.Service }}
-    app.kubernetes.io/instance: {{ .Release.Name }}
+    app.kubernetes.io/instance: {{ .Values.clusterName }}
     app.kubernetes.io/component: importer
     helm.sh/chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+"  "_" }}
 spec:
-  clusterIP: None
```
Comment (DanielZhangQD) [inline charts/tikv-importer/templates/tikv-importer-service.yaml:12]: We should keep this, why remove it here?
This service is a headless service through which Prometheus can access the Pod via Pod name.

**Comment by DanielZhangQD** [comment]
LGTM cancel

File: charts/tikv-importer/templates/tikv-importer-service.yaml (hunk 5 above)
Comment (csuzhangxc) [reply charts/tikv-importer/templates/tikv-importer-service.yaml:12]: reverted.

**Review by DanielZhangQD — APPROVED** [review]
LGTM
